### PR TITLE
incorporate relevant modules from MAP pipeline

### PR DESCRIPTION
**Pull request by ttngu207**

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2017 Mesoscale Activity Map
```
Comment (vncntprvst) [inline LICENSE]: Shouldn't we change the project name here?

File: pipeline/__init__.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+# safe-guard in case `custom` is not provided
+if 'custom' not in dj.config:
+    dj.config['custom'] = {}
+
+
+def get_schema_name(name):
+    try:
+        return dj.config['custom']['{}.database'.format(name)]
+    except KeyError:
+        prefix = 'orofacial_'
```
Comment (vncntprvst) [inline pipeline/__init__.py]: change prefix to cosmo_?

File: pipeline/ephys.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    unit_quality_description :  varchar(4000)
+    """
+    contents = [
+        ('good', 'single unit'),
+        ('ok', 'probably a single unit, but could be contaminated'),
+        ('multi', 'multi unit'),
+        ('all', 'all units')
+    ]
+
+
```
Comment (vncntprvst) [inline pipeline/ephys.py:91]: We need to keep track of whether units were phototagged (responsive to light). That could go in a schema like this, maybe:
```suggestion
@schema
class Phototag(dj.Manual):
    definition = """
    -> PhotoStim
    ---
    responses  : varchar(30)   # Yes / No / Maybe
    responsive_channels= null : varchar(30)  # responsive channels
    response_delay=  null : smallint # ms
    """
```

File: pipeline/ephys.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    cell_type  :  varchar(100)
+    ---
+    cell_type_description :  varchar(4000)
+    """
+    contents = [
+        ('Pyr', 'putative pyramidal'),
+        ('FS', 'fast spiking'),
+        ('not classified', 'intermediate spike-width that falls between spike-width thresholds for FS or Putative pyramidal cells'),
+        ('all', 'all types')
+    ]
```
Comment (vncntprvst) [inline pipeline/ephys.py]: This project is focused on brainstem circuits, so you can remove 'Pyr', 'putative pyramidal'. 
Keep 'FS', 'fast spiking'
Keep 'not classified', but remove the text description 
Then, we should add cell types as classified from phototagging:
'PM', 'premotor'
'Proj', 'projection neuron'
'NTNG1', 'Netrin-G1'
'PV', 'Parvalbumin positive neuron'
I'll need to add more labels as time goes. How easy would it be to do that? 
Also, is there a way to apply several labels, or should we create one label for each combination (e.g., 'PV PM')?

File: pipeline/ephys.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+
+@schema
+class ClusteringMethod(dj.Lookup):
+    definition = """
+    clustering_method: varchar(16)
+    """
+
+    contents = zip(['jrclust_v3', 'kilosort', 'jrclust_v4', 'kilosort2'])
```
Comment (vncntprvst) [inline pipeline/ephys.py:117]: Add 'other' to content (we'll need to revisit that)

File: pipeline/ephys.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        self.insert1(dict(key, cell_type='FS' if waveform_width < 0.4 else 'Pyr'))
+
+
+@schema
+class UnitStat(dj.Computed):
+    definition = """
+    -> Unit
+    ---
+    isi_violation=null: float    # 
+    avg_firing_rate=null: float  # (Hz)
```
Comment (vncntprvst) [inline pipeline/ephys.py:220]:
```suggestion
    avg_firing_rate=null: float  # (Hz)
    avg_cv2=null: float  #
```

File: pipeline/experiment.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    ----
+    task_description : varchar(4000)
+    """
+    contents = [
+        ('audio delay', 'auditory delayed response task (2AFC)'),
+        ('audio mem', 'auditory working memory task'),
+        ('s1 stim', 'S1 photostimulation task (2AFC)'),
+        ('foraging', 'foraging task based on Bari-Cohen 2019'),
+        ('foraging 3lp', 'foraging task based on Bari-Cohen 2019 with variable delay period')
+    ]
```
Comment (vncntprvst) [inline pipeline/experiment.py]: replace contents with 
```suggestion
    contents = [
        ('fm texture discrim', 'freely moving texture discrimination)'),
        ('hf texture discrim', 'head fixed texture discrimination)'),
        ('hf pole loc', 'head fixed pole localization)'),
        ('hf wall dist', 'head-fixed wall distance'),
        ('hf wheel', 'head-fixed wheel runing')
        ('gen tun', 'unstructured generic tuning stimulation)')
    ]
```
@ksseverson57 For Wenxi's task, does this work ?  ('hf wall dist', 'head-fixed wall distance')
other task suggestions?

File: pipeline/experiment.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+        ('s1 stim', 3, 'full distractors, with 2 distractors (at different times) on some of the left trials'),
+        ('s1 stim', 4, 'full distractors'),
+        ('s1 stim', 5, 'mini-distractors, with different levels of the mini-stim during sample period'),
+        ('s1 stim', 6, 'full distractors; same as protocol 4 but with a no-chirp trial-type'),
+        ('s1 stim', 7, 'mini-distractors and full distractors (only at late delay)'),
+        ('s1 stim', 8, 'mini-distractors and full distractors (only at late delay), with different levels of the mini-stim and the full-stim during sample period'),
+        ('s1 stim', 9, 'mini-distractors and full distractors (only at late delay), with different levels of the mini-stim and the full-stim during sample period'),
+        ('foraging', 100, 'moving lickports, delay period, early lick punishment, sound GO cue then free choice'),
+        ('foraging 3lp', 101, 'moving lickports, delay period, early lick punishment, sound GO cue then free choice from three lickports')
+    ]
```
Comment (vncntprvst) [inline pipeline/experiment.py]: This will need to change, once we have settled on Task types above.

File: pipeline/experiment.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+            lat = 'both'
+        elif (ml_locations > 0).all():
+            lat = 'right'
+        elif (ml_locations < 0).all():
+            lat = 'left'
+        else:
+            assert (ml_locations == 0).all()  # sanity check
+            raise ValueError('Ambiguous hemisphere: ML locations are all 0...')
+
+        self.insert1(dict(key, stim_brain_area=brain_areas[0], stim_laterality=lat))
```
Comment (vncntprvst) [inline pipeline/experiment.py:107]: We'll duplicate and modify this for electrical stimulations

File: pipeline/experiment.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+    # Mouse training
+    training_type : varchar(100) # mouse training
+    ---
+    training_type_description : varchar(2000) # description
+    """
+    contents = [
+         ('regular', ''),
+         ('regular + distractor', 'mice were first trained on the regular S1 photostimulation task  without distractors, then the training continued in the presence of distractors'),
+         ('regular or regular + distractor', 'includes both training options')
+         ]
```
Comment (vncntprvst) [inline pipeline/experiment.py]: To be changed

File: pipeline/experiment.py [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+    """
+    contents = zip(('left', 'right', 'middle', 'none'))
+
+
+@schema
+class Outcome(dj.Lookup):
+    definition = """
+    outcome : varchar(32)
+    """
+    contents = zip(('hit', 'miss', 'ignore'))
```
Comment (vncntprvst) [inline pipeline/experiment.py:204]: why not 'false alarm' and 'correct rejection' ?

File: pipeline/experiment.py [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+        depth:       decimal(6, 2) # (um) manipulator depth relative to surface of the brain (0); more ventral is more negative
+        theta:       decimal(5, 2) # (deg) - elevation - rotation about the ml-axis [0, 180] - w.r.t the z+ axis
+        phi:         decimal(5, 2) # (deg) - azimuth - rotation about the dv-axis [0, 360] - w.r.t the x+ axis
+        ---
+        -> lab.BrainArea           # target brain area for photostim 
+        """
+
+
+@schema
+class PhotostimBrainRegion(dj.Computed):
```
Comment (vncntprvst) [inline pipeline/experiment.py:84]: We also do external (skin) photostimulations, so we could rename this class `PhotostimRegion`

File: pipeline/lab.py [hunk 13]
```diff
@@ -0,0 +1,409 @@
+import datajoint as dj
+import numpy as np
+from . import get_schema_name
+
+schema = dj.schema(get_schema_name('lab'))
+
+
```
Comment (vncntprvst) [inline pipeline/lab.py:7]: External collaborators from other labs (esp. UCSD) may contribute data. 
Add 
```suggestion
@schema
class Lab(dj.Manual): 
    definition = """ # Lab
    lab : varchar(255)  #  lab conducting the study
    ----
    institution  : varchar(255)  # Institution to which the lab belongs
    """
```

File: pipeline/lab.py [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+    """
+    contents = zip(['pl56', 'kj18'])
+
+
+@schema
+class AnimalSource(dj.Lookup):
+    definition = """
+    animal_source       : varchar(30)
+    """
+    contents = zip(['Jackson Labs', 'Allen Institute', 'Charles River', 'MMRRC', 'Taconic', 'Other'])
```
Comment (vncntprvst) [inline pipeline/lab.py]:
```suggestion
    contents = zip(['Jackson Labs', 'Allen Institute', 'Charles River', 'MMRRC', 'Taconic', 'Lab made', 'Other'])
```

File: pipeline/lab.py [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+
+
+# ---- Animal information ----
+
+@schema
+class AnimalStrain(dj.Lookup):
+    definition = """
+    animal_strain       : varchar(30)
+    """
+    contents = zip(['pl56', 'kj18'])
```
Comment (vncntprvst) [inline pipeline/lab.py]:
```suggestion
    contents = zip(['C57BL6', 'Ai14', 'Ai32', 'Ai35', 'Ai65D',
        'Emx1_Cre', 'GAD2_Cre', 'vGLut2_Cre', 'Pv_Cre',
        'Pv_CreERt2', 'Pv_CreN', 'TrpV1_Cre', 'Netrin_G1_Cre',
        'FosTVA', 'Rphi_AP', 'Rphi_tomato', 'Rphi_GFP', 'ChodlPLAP'])
```

File: pipeline/lab.py [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+    """
+
+
+@schema
+class Subject(dj.Manual):
+    definition = """
+    subject_id          : int   # institution 6 digit animal ID
+    ---
+    -> [nullable] Person        # person responsible for the animal
+    cage_number         : int   # institution 6 digit animal ID
```
Comment (vncntprvst) [inline pipeline/lab.py]:
```suggestion
    cage_number         : int   # institution 7 digit animal ID
```

File: pipeline/lab.py [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+    gene_modification   : varchar(60)
+    ---
+    gene_modification_description = ''         : varchar(256)
+    """
+
+
+@schema
+class Subject(dj.Manual):
+    definition = """
+    subject_id          : int   # institution 6 digit animal ID
```
Comment (vncntprvst) [inline pipeline/lab.py]: We often mix letters and numbers to identify subjects 
```suggestion
    subject_id          : varchar(10)  # 
```

File: pipeline/lab.py [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+
+
+# ---- Virus information ----
+
+@schema
+class VirusSource(dj.Lookup):
+    definition = """
+    virus_source   : varchar(60)
+    """
+    contents = zip(['Janelia', 'UPenn', 'Addgene', 'UNC', 'Other'])
```
Comment (vncntprvst) [inline pipeline/lab.py]:
```suggestion
    contents = zip(['Janelia', 'UPenn', 'Addgene', 'UNC', 'MIT', 'Duke', 'Lab made', 'Other'])
```

File: pipeline/lab.py [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+
+
+@schema
+class Serotype(dj.Manual):
+    definition = """
+    serotype   : varchar(60)
+    """
+
```
Comment (vncntprvst) [inline pipeline/lab.py:138]:
```suggestion
# For nomenclature reference, here are some vectors/compounds used in the lab
# 'Dextran', 'CTb', 'AAV', 'AAV2', 'AAV2_1',
#        'AAV2_5', 'AAV2_8', 'AAV8', 'AAV2_9', 'AAV9', 'AAV',
#        'retroAAV', 'LV', 'FuGB2_LV', 'RG_LV', 'CANE_LV',
#        'EnVA_SAD_dG_RV', 'RG_CVS_N2cdG_RV', 'CANE_RV', 'KainicAcid'

# Nomenclature reference for constructs
# 'Alexa568', 'TMR', 'GFP', 'EGFP', 'mNeonG',
#        'tdTomato', 'mCherry', 'RFP', 'Cre', 'Syn_Cre', 'CreC',
#        'EF1a_mCherry_IRES_WGA_Cre', 'EF1a_Flex_ChR2',
#        'Flex_TVAmCherry', 'Flex_TVA_RG_GFP'
```

File: pipeline/lab.py [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+        x_coord=NULL: float  # (um) x coordinate of the electrode within the probe, (0, 0) is the bottom left corner of the probe
+        y_coord=NULL: float  # (um) y coordinate of the electrode within the probe, (0, 0) is the bottom left corner of the probe
+        z_coord=0: float     # (um) z coordinate of the electrode within the probe, (0, 0) is the bottom left corner of the probe
+        """
+
+    @property
+    def contents(self):
+        return zip(['silicon_probe', 'tetrode_array',
+                    'neuropixels 1.0 - 3A', 'neuropixels 1.0 - 3B',
+                    'neuropixels 2.0 - SS', 'neuropixels 2.0 - MS'])
```
Comment (vncntprvst) [inline pipeline/lab.py]:
```suggestion
        return zip(['silicon_probe', 'tetrode_array', 'single electrode',
                    'neuropixels 1.0 - 3A', 'neuropixels 1.0 - 3B',
                    'neuropixels 2.0 - SS', 'neuropixels 2.0 - MS'])
```

File: pipeline/lab.py [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+    definition = """
+    photostim_device  : varchar(20)
+    ---
+    excitation_wavelength :  decimal(5,1)  # (nm) 
+    photostim_device_description : varchar(255)
+    """
+    contents =[
+       ('LaserGem473', 473, 'Laser (Laser Quantum, Gem 473)'),
+       ('LED470', 470, 'LED (Thor Labs, M470F3 - 470 nm, 17.2 mW (Min) Fiber-Coupled LED)'),
+       ('OBIS470', 473, 'OBIS 473nm LX 50mW Laser System: Fiber Pigtail (Coherent Inc)')]
```
Comment (vncntprvst) [inline pipeline/lab.py:426]: Will update that.

File: pipeline/psth.py [hunk 22]
```diff
@@ -0,0 +1,579 @@
...
+    trial_condition_hash:       varchar(32)     # trial condition hash - hash of func and arg
+    unique index (trial_condition_hash)
+    trial_condition_func:       varchar(36)     # trial retrieval function
+    trial_condition_arg:        longblob        # trial retrieval arguments
+    """
+
+    @property
+    def contents(self):
+        contents_data = [
+            {
```
Comment (vncntprvst) [inline pipeline/psth.py]: @ksseverson57 , can you see with Wenxi what PSTH she'd want for her task?

File: pipeline/psth.py [hunk 23]
```diff
@@ -0,0 +1,579 @@
...
+                'trial_condition_func': '_get_trials_exclude_stim',
+                'trial_condition_arg': {
+                    '_outcome': 'ignore',
+                    'task': 'audio delay',
+                    'task_protocol': 1,
+                    'early_lick': 'no early',
+                    'auto_water': 0,
+                    'free_water': 0,
+                    'trial_instruction': 'right'}
+            }
```
Comment (vncntprvst) [inline pipeline/psth.py]: We do not have an audio delay task
```suggestion
            }
```

File: pipeline/psth.py (hunk 22 above)
Comment (vncntprvst) [reply pipeline/psth.py]:
```suggestion
    def contents(self):
        contents_data = [
            {   'trial_condition_name': 'touch_protraction',
                'trial_condition_func': '_get_trials_exclude_stim',
                'trial_condition_arg': {
                    'task': 'hf pole loc',
                    'task_protocol': 1,
                    'outcome': 'hit',
                    'early_lick': 'no early',
                    'auto_water': 0,
                    'free_water': 0}
            },
            {   'trial_condition_name': 'touch_retraction',
                'trial_condition_func': '_get_trials_exclude_stim',
                'trial_condition_arg': {
                    'task': 'hf pole loc',
                    'task_protocol': 1,
                    'outcome': 'hit',
                    'early_lick': 'no early',
                    'auto_water': 0,
                    'free_water': 0}
            }
```
Need to add similar conditions for other tasks

File: pipeline/tracking.py [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+    tracking_device:                    varchar(20)     # device type/function
+    ---
+    tracking_position:                  varchar(20)     # device position
+    sampling_rate:                      decimal(8, 4)   # sampling rate (Hz)
+    tracking_device_description:        varchar(100)    # device description
+    """
+    contents = [
+        ('Camera 0', 'side', 1/0.0034, 'Chameleon3 CM3-U3-13Y3M-CS (FLIR)'),
+        ('Camera 1', 'bottom', 1/0.0034, 'Chameleon3 CM3-U3-13Y3M-CS (FLIR)'),
+        ('Camera 2', 'body', 1/0.01, 'Chameleon3 CM3-U3-13Y3M-CS (FLIR)')]
```
Comment (vncntprvst) [inline pipeline/tracking.py:27]: To be changed

File: pipeline/tracking.py [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+
+    class WhiskerTracking(dj.Part):
+        definition = """
+        -> Tracking
+        whisker_idx:          int             # 0, 1, 2
+        ---
+        whisker_x:            longblob        # whisker x location (px) over time
+        whisker_y:            longblob        # whisker y location (px) over time
+        face_angle:           longblob
+        whikser_length:       longblob
```
Comment (vncntprvst) [inline pipeline/tracking.py]:
```suggestion
        length:       longblob # wisker length (px) 
```

File: pipeline/lab.py [hunk 26]
```diff
@@ -0,0 +1,409 @@
...
+
+
+@schema
+class Whisker(dj.Lookup):
+    definition = """
+    whisker: varchar(32)
+    """
+    contents = zip(['B1', 'B2', 'B3',
+                    'C1', 'C2', 'C3',
+                    'D1', 'D2', 'D3'])
```
Comment (vncntprvst) [inline pipeline/lab.py]:
```suggestion
    contents = zip(['Alpha','A1','A2','A3','A4',
                    'Beta','B1', 'B2', 'B3','B4',
                    'Gamma','C1', 'C2', 'C3','C4', 'C5', 'C6',
                    'Delta','D1', 'D2', 'D3','D4', 'D5', 'D6'])
```

Comment (vncntprvst) [reply pipeline/lab.py]: Could also add non-whisker pad ones ...

File: pipeline/tracking.py [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+        right_paw_likelihood:   longblob        # right_paw location likelihood
+        """
+
+    class WhiskerTracking(dj.Part):
+        definition = """
+        -> Tracking
+        whisker_idx:          int             # 0, 1, 2
+        ---
+        whisker_x:            longblob        # whisker x location (px) over time
+        whisker_y:            longblob        # whisker y location (px) over time
```
Comment (vncntprvst) [inline pipeline/tracking.py]: That's intended to be the whisker's pixel contour, right?

File: pipeline/tracking.py [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+        face_angle:           longblob
```
Comment (vncntprvst) [inline pipeline/tracking.py]:
```suggestion
        angle:           longblob # mean angle at follicle
        curvature:     longblob # mean curvature (1/mm)
        face_x:           longblob # approximate center of whisker pad, x (px)
        face_y:           longblob # approximate center of whisker pad, y (px) 
        follicle_x:           longblob #follicle position: x (px)
        follicle_y:           longblob # follicle position: y (px)
        tip_x:           longblob # tip position: x (px)
        tip_y:           longblob # tip position: y (px)
```

**Review by vncntprvst — COMMENTED** [review]
Here's a first pass, it will need a few more modifications

File: pipeline/ephys.py (hunk 4 above)
Comment (ttngu207) [reply pipeline/ephys.py]: adding new values to the lookup table is no problem at all, that is precisely the intended usage for lookup table.

The choice of one-to-one or multiple labels for each cell is at the `UnitCellType` table. The current definition enforces one-to-one:
```
@schema
class UnitCellType(dj.Computed):
    definition = """
    -> Unit
    ---
    -> CellType
    """
```

To accommodate multiple cell-type labels, then this design would work:
```
@schema
class UnitCellType(dj.Computed):
    definition = """
    -> Unit
    -> CellType
    """
```

File: pipeline/ephys.py (hunk 5 above)
Comment (ttngu207) [reply pipeline/ephys.py:117]: Contents in lookup table are not fixed, so we can add new clustering method as new method comes up.

File: pipeline/psth.py (hunk 23 above)
Comment (ttngu207) [reply pipeline/psth.py]: the content (and perhaps usage) of this table will need to be revisited

File: pipeline/ephys.py [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+
+@schema
+class Clustering(dj.Imported):
+    definition = """
+    -> ProbeInsertion
+    -> ClusteringMethod
+    ---
+    clustering_time: datetime  # time of generation of this set of clustering results 
+    quality_control: bool  # has this clustering results undergone quality control
+    manual_curation: bool  # is manual curation performed on this clustering result
```
Comment (vncntprvst) [inline pipeline/ephys.py:128]: Currently I'm using KS2 for clustering and JRClust for manual curation. How should we record this?

File: pipeline/ephys.py [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+        """
+
+
+@schema
+class UnitCellType(dj.Computed):
+    definition = """
+    -> Unit
+    ---
+    -> CellType
+    """
```
Comment (vncntprvst) [inline pipeline/ephys.py:193]:
```suggestion
@schema
class UnitCellType(dj.Computed):
    definition = """
    -> Unit
    -> CellType
    """
```

File: pipeline/psth.py (hunk 23 above)
Comment (vncntprvst) [reply pipeline/psth.py]: @xiaow2 could you write down a quick description of PSTHs that you'd like to plot for your task?  Thanks

File: pipeline/ephys.py (hunk 4 above)
Comment (vncntprvst) [reply pipeline/ephys.py]: OK!

**Pull request merged**